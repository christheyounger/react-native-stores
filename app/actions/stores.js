import {
    GET_STORES,
    GOT_STORES,
    GET_STORE,
    GOT_STORE,
    GET_STORES_FAIL
} from '../constants/actiontypes';

export const getStores = () => ({
    type: GET_STORES,
});

export const gotStores = stores => ({
    type: GOT_STORES,
    payload: { stores },
});

export const getStoresFail = error => ({
    type: GET_STORES_FAIL,
    payload: { error },
})

export const getStore = () => ({
    type: GET_STORE,
});

export const gotStore = store => ({
    type: GOT_STORE,
    payload: { store },
});
