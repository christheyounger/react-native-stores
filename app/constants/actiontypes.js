export const GET_STORES = "GET_STORES";
export const GOT_STORES = "GOT_STORES";
export const GET_STORES_FAIL = "GET_STORES_FAIL";
export const GET_STORE = "GET_STORE";
export const GOT_STORE = "GOT_STORE";
export const GET_STORE_FAIL = "GET_STORE_FAIL";
export const GET_STORE_BRANCHES = "GET_STORE_BRANCHES";
export const GOT_STORE_BRANCHES = "GOT_STORE_BRANCHES";
export const GET_STORE_BRANCHES_FAIL = "GET_STORE_BRANCHES_FAIL";
