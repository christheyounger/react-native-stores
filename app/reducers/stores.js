import {
    GET_STORES,
    GOT_STORES,
} from '../constants/actiontypes';

const initialState = {
    loading: true,
    stores: []
};

export default function stores(state = initialState, action) {
    switch(action.type) {
        case GET_STORES:
            return initialState;
        case GOT_STORES:
            return {...state, stores: action.payload.stores, loading: false};
        default:
            return state;
    }
}
