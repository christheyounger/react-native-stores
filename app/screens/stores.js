import React from 'react';
import Store from '../components/Store';
import { FlatList, ActivityIndicator, View  } from 'react-native';

export default class StoresList extends React.Component {
  constructor(props) {
    super(props);
  }

  componentDidMount(){
    this.props.getStores();
  }

  handleClick(item) {
    this.props.navigation.navigate('StoreDetails', item);
  }

  render(){
    if(this.props.loading){
      return(
        <View style={{flex: 1, padding: 20}}>
          <ActivityIndicator/>
        </View>
      )
    }
  

    return(
      <View>
        <FlatList
          data={this.props.stores}
          renderItem={({item}) =>
            <Store
              id={item.id}
              name={item.name}
              branches={item.branches}
              handleClick={item => this.handleClick(item)}
            />}
          keyExtractor={({id}) => id.toString()}
        />
      </View>
    );
  }
}
