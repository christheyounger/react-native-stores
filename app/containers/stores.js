import { getStores, gotStores } from "../actions/stores";
import { connect } from 'react-redux';
import StoresList from '../screens/stores';

const matchStateToProps = (state) => {
    const { stores: { loading, stores} } = state;
    return { loading, stores };
};

const mapDispatchToProps = dispatch => ({
    getStores: () => dispatch(getStores()),
    gotStores: payload => dispatch(gotStores(payload)),
});

export default connect(matchStateToProps, mapDispatchToProps)(StoresList);
