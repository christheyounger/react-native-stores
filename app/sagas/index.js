import { all } from 'redux-saga/effects';
import { watchGetStores } from './stores';

export default function* rootSaga() {
    yield all([
        watchGetStores(),
    ])
}
