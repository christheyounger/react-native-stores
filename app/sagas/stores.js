import { put, call, takeLatest } from 'redux-saga/effects'
import settings from '../config/settings';
import { gotStores, getStoresFail } from '../actions/stores';
import { handleErrors } from '../utilities'

export function* getStores() {
    try {
        const response = yield call(fetch, settings.apiUrl + 'stores')
        handleErrors(response);
        const stores = response.json();
        yield put(gotStores(stores));
    } catch (error) {
        yield put(getStoresFail(error))
    }
}

export function* watchGetStores() {
    yield takeLatest('GET_STORES', getStores)
}