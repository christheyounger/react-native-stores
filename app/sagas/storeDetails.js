import { all, put, call, takeLatest } from 'redux-saga/effects'
import settings from "../config/settings";
import { handleErrors } from "../utilities";
import { 
    gotStore,
    getStoreFail,
    getStoreBranches,
    gotStoreBranches,
    getStoreBranchesFail
} from "../actions/stores";
import { GET_STORE, GET_STORE_BRANCHES } from '../constants/actiontypes';

export function* getStore(id) {
    try {
        const response = yield call(fetch, settings.apiUrl + 'store/' + id);
        handleErrors(response);
        const store = response.json();
        yield all([put(gotStore(store)), put(getStoreBranches(store.id))]);
    } catch (error) {
        yield put(getStoreFail(error));
    }
}

export function* watchGetStore() {
    yield takeLatest(GET_STORE, getStore);
}

export function* getStoreBranchesSaga(id) {
    try {
        const response = yield call(fetch, `${settings.apiUrl}'store/${id}/branches`)
        handleErrors(response);
        const branches = response.json();
        yield put(gotStoreBranches(branches));
    } catch (error) {
        yield put(getStoreBranchesFail(error));
    }
}

export function* watchGetBranches() {
    yield takeLatest(GET_STORE_BRANCHES, getStoreBranchesSaga);
}
