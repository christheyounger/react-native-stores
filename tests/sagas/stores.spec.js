import 'isomorphic-fetch';
import { getStores, watchGetStores } from "../../app/sagas/stores";
import { runSaga } from 'redux-saga';
import { takeLatest } from 'redux-saga/effects';
import { GET_STORES } from '../../app/constants/actiontypes';

describe('getStores', () => {
  const StoresList = [{id: 1, name: 'new store 1'}];
  test('should handle success', async () => {
    const dispatched = [];
    const dispatch = (action) => { dispatched.push(action) };

    const mockPromise = Promise.resolve({
        ok: true,
        json: () => StoresList,
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockPromise);
    const stores = await runSaga({
        dispatch,
        getState: () => ({loading: false})
    }, getStores);
    expect(global.fetch).toMatchSnapshot();
    expect(dispatched).toMatchSnapshot();
    global.fetch.mockClear();
  });
  test('should handle failure', async () => {
    const dispatched = [];
    const dispatch = (action) => { dispatched.push(action) };

    const mockPromise = Promise.resolve({
        ok: false,
        statusText: '404 Not Found',
    });
    jest.spyOn(global, 'fetch').mockImplementation(() => mockPromise);
    const stores = await runSaga({
        dispatch,
        getState: () => ({loading: false})
    }, getStores);
    expect(global.fetch).toMatchSnapshot();
    expect(dispatched).toMatchSnapshot();
    global.fetch.mockClear();
  });
  test('should wait for GET_STORES', () => {
    const saga = watchGetStores();
    expect(saga.next().value).toEqual(takeLatest(GET_STORES, getStores))
  });
});
