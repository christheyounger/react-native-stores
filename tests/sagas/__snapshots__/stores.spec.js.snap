// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStores should handle failure 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:8000/api/stores",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "_40": 0,
        "_55": Object {
          "ok": false,
          "statusText": "404 Not Found",
        },
        "_65": 1,
        "_72": null,
      },
    },
  ],
}
`;

exports[`getStores should handle failure 2`] = `
Array [
  Object {
    "payload": Object {
      "error": [Error: 404 Not Found],
    },
    "type": "GET_STORES_FAIL",
  },
]
`;

exports[`getStores should handle success 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:8000/api/stores",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {
        "_40": 0,
        "_55": Object {
          "json": [Function],
          "ok": true,
        },
        "_65": 1,
        "_72": null,
      },
    },
  ],
}
`;

exports[`getStores should handle success 2`] = `
Array [
  Object {
    "payload": Object {
      "stores": Array [
        Object {
          "id": 1,
          "name": "new store 1",
        },
      ],
    },
    "type": "GOT_STORES",
  },
]
`;
